### ISSUE TITLE
Document size limitation for variables (from ES: 32766 bytes)

### ISSUE THREAD

**Issue by berndruecker** [zeebe-io/zeebe#3259]
As variables are indexed in ES for operate this poses a limitation in their size, otherwise you end up with the error below.

We should document maximum variable sizes (and if this limit is configurable in ES probably also hint to this).

```
[exporter] [172.25.0.2:26501-zb-fs-workers-0] WARN  io.zeebe.broker.exporter.elasticsearch - Failed to flush at least one bulk request ElasticsearchException[Elasticsearch exception [type=illegal_argument_exception, reason=Document contains at least one immense term in field="value.value" (whose UTF8 encoding is longer than the max length 32766), all of which were skipped.  Please correct the analyzer to not produce such terms.  The prefix of the first immense term is: '[91, 123, 34, 95, 105, 100, 34, 58, 34, 53, 100, 98, 49, 52, 99, 97, 55, 49, 53, 50, 55, 102, 99, 97, 49, 48, 49, 100, 50, 50]...', original message: bytes can be at most 32766 in length; got 37546]]; nested: ElasticsearchException[Elasticsearch exception [type=max_bytes_length_exceeded_exception, reason=bytes can be at most 32766 in length; got 37546]]
```

**Comment by jwulf** [comment]
Is this limit per variable, or for the entire payload of variables?

**Comment by MrSaints** [comment]
This is occurring more frequently, especially when using call activities:

```
│ 2020-01-29 14:41:37.258 [Broker-0-Exporter-1] [Broker-0-zb-fs-workers-0] WARN  io.zeebe.broker.exporter.elasticsearch - Failed to flush at least one bulk request ElasticsearchException[Elasticsearch exception [type=illegal_argument_exception, reason=Document contains at least one immense term in field="value.value" (whose UTF8 encoding is longer than the m │
│ ax length 32766), all of which were skipped.  Please correct the analyzer to not produce such terms.  The prefix of the first immense term is: '[91, 34, 99, 107, 53, 122, 101, 101, 98, 122, 52, 48, 48, 48, 51, 48, 49, 99, 50, 100, 101, 111, 54, 54, 110, 114, 111, 34, 44, 34]...', original message: bytes can be at most 32766 in length; got 47013]]; nested:  │
│ ElasticsearchException[Elasticsearch exception [type=max_bytes_length_exceeded_exception, reason=bytes can be at most 32766 in length; got 47013]];                                                                                                                                                                                                                    │
```

The annoying this is, this pretty much breaks the data consistency between "what Operate knows" versus "what is happening, and what has happened in Zeebe".

**Comment by strawhat5** [comment]
Hi @berndruecker , I guess we have 2 potential solutions to this issue:

**Solution 1:** Applying [`ignore_above`](https://www.elastic.co/guide/en/elasticsearch/reference/current/ignore-above.html) property to [`variables`](https://github.com/zeebe-io/zeebe/blob/develop/exporters/elasticsearch-exporter/src/main/resources/zeebe-record-workflow-instance-creation-template.json#L27) mappings in all the Elastic exporter [templates](https://github.com/zeebe-io/zeebe/tree/develop/exporters/elasticsearch-exporter/src/main/resources) can mitigate this issue. This would need a code change and new release of zeebe-elastic-exporter.

**Solution 2:** Clients validate byte size of all variables, and trim them into multiple variables of size<32KB.

I would be happy to hear your thoughts on which should be the preferable way to go forward.

**Comment by jwulf** [comment]
Option 3 could be refactor to store large objects in a layer outside Zeebe (Redis/S3) and pass around references to them.

I'm not sure what impact large variables has on memory and disk usage, but I am pretty sure it will have an effect on end-to-end latency because of the need to replicate events before processing them. 

If you are using fetchVariables to filter on activateJobs calls it will be mitigated, but I would load test it and compare 1000s of runs with dummy 32k payloads against runs with link references. 

There will be a trade-off on calls to the cacHing layer for each worker, plus the provisioning and management of the caching layer, but you'd want to quantify the cost and the benefit.

**Comment by berndruecker** [comment]
Option 3 is a best practice, that make total sense (see also https://camunda.com/best-practices/handling-data-in-processes/), but for the core workflow engine we can't rely on users doing it "right" - so it does not really solve the issue.

Option 2 seems like a rabbit hole - as we have many client libraries, some of them developed by brave community members. So it will be hard to maintain some consistency across them.

Option 1 for me seems to be a good approach, as it would simply ignore the variable in the export (as then *documented* :-)), but the core engine will work as expected.

**Comment by strawhat5** [comment]
Thanks for your suggestions @jwulf @berndruecker !

I guess we will go with Option 3 for now, as the next Zeebe release might take some time. But as Bernd correctly pointed that it is just a best practice, and the actual solution would be Option 1; we would wait for Option 1 to be integrated in the next release (hopefully 🙂).

**Comment by menski** [comment]
closed with #3891
